Userland/Libraries/LibTest/Randomized/README.md

```diff
@@ -0,0 +1,119 @@
+the "integrated shrinking" lazy-tree based implementations (Hedgehog etc.) or
+the manual shrinking ones (original QuickCheck etc.). (More information in this
+talk: https://www.youtube.com/watch?v=WE5bmt0zBxg)
+
+## Implementation
+
+On a very high level, the PBT runner remembers the random bits (`RandomRun`)
+used when generating the random value, and when it finds a value that fails the
+user's test, instead of trying to shrink the generated value it tries to shrink
+these random bits and then generate a new value from them. 
+
+This makes shrinking work automatically on any data you might want to generate.
+
+Generators are implemented in such a way that a smaller/shorter `RandomRun` will
+result in a simpler generated value.
+
+If the new RandomRun can be successfully turned into a value and if the test
+still fails on that new simpler value, we keep the RandomRun as our new best and
+try again. Once we can't improve things anymore, we turn on error reporting, run
+the test one last time with the final RandomRun, and report the failing value to
+the user.
+
+## Code organization
+
+- `TestResult.h`
+  - Defines an enum class TestResult.
+    This expands the typical "passed / failed": we also need to care about
+    a generator rejecting a RandomRun (eg. when the user calls the ASSUME(...)
+    macro with a predicate that can't be satisfied).
+
+- `Generator.h`
+  - Contains generators: fns of shape T(), eg. `u32 Gen::unsigned_int(u32 max)`
+    - These implicitly depend on a RandomnessSource held by the singleton
+      TestSuite.
+  - These can be called directly, but the top-level use by the user should always
+    happen via the GEN(...) macro which makes sure the generated value gets
+    logged to the user in case of a failure.
+  - Example:
+    `Gen::vector(1, 4, []() { return Gen::unsigned_int(5); })`
+    generates vectors of length between 1 and 4, of unsigned ints in range 0..5.
+    Eg. `{2,5,3}`, `{0}`, `{1,5,5,2}`.
+
+- `RandomnessSource.h`
+  - A source of random bits.
+  - There are two variants of `RandomnessSource`:
+    - Live: gives `AK/Random` u32 values and remembers them into a `RandomRun`
+    - Recorded: gives (replays) u32 values from a static `RandomRun`
+
+- `RandomRun.h`
+  - A finite sequence of random bits (in practice, `u32`s).
+  - Example: `{2,5,0,11,8,0,0,1}`
+
+- `ShrinkCommand.h`
+  - A high-level recipe for how to try and minimize a given `RandomRun`.
+  - For example, "zero this contiguous chunk of it" or "minimize the number on
+    this index using binary search".
+  - These later get interpreted by the PBT runner on a specific `RandomRun`.
+
+- `Chunk.h`
+  - A description of a contiguous `RandomRun` slice.
+  - Example: `Chunk{size = 4, index = 2}`:  [_,_,X,X,X,X,...]
+
+- `Shrink.h`
+  - Algorithms for interpreting `ShrinkCommand`s and the main shrinking loop
+
+- `TestCase.h`
+  - The `TestCase::randomized(...)` function contains the main testing loop
```

Comment:
While we appreciate the effort, to me it looks like the comments within the actual files will be more than sufficient for explaining the structure - duplicating the information here will probably just open up opportunities for the documentation to be outdated, and they are probably closer to internals anyways.

Reply:
If we really want the file, it could live in Documentation/ .

Reply:
I felt like a summary "all in one place" would be helpful to get an overall picture, but I can remove it if you don't feel like it helped you much.

I was also contemplating another file with more hands-on examples of _using_ this library, explaining the macros and so on. Examples of writing generators, writing the tests, etc... WDYT?

Reply:
Personally, I don't mind the "Example", "Property based testing?" and "Implementation" parts, but as Andrew already pointed out, they should be in `Documentation` (or maybe the manpages?) instead.

What I was talking about specifically is the "Code organization" part. We try to write nicely organized code to begin with (and encourage developers to just look up questions there by using the go-to-definition feature of their IDE), so we usually don't keep around a separate text that explains the code in particular. If there is any information in there that isn't reflected anywhere else already, it should probably be inlined either into the code comments or into the sections located above.